### ISSUE TITLE
backtesting with timeframe-detail fills orders on "long" timeframe

### ISSUE THREAD

**Issue by mciepluc** [freqtrade/freqtrade#7529]
## Describe your environment

  * Operating system: All
  * Python Version: 3.10
  * CCXT version: 1.93.90
  * Freqtrade Version: 2022.9

## Describe the problem:

Seems there is incorrect handling of the custom_stoploss in backtesting. When timeframe detail used, the stoploss call does not match the expected price changes within the timeframe.
Example:
- timeframe=1h, timeframe-detail=1m
- enter trade and stop loss transactions will happen at a single 1h candle (let's say, candle starts at 0:00)
- enter trade transaction is in the middle of the candle (e.g. 0:45)
- custom_stoploss seems to be activated before actual enter trade, so seems this function does not respect the timeframe-detail parameter (assume stop loss condition was satisfied only between 0:30-0:35)
- observed trade executed at the price from 0:45 and stop loss price taken from 0:30

### Steps to reproduce:

  1. request a transaction that will be executed in the middle of the candle
  2. define a custom-stoploss,
  3. run a backtesting with timeframe-detail
  4. incorrect stop loss transaction may be observed when stop loss conditions satisfied at the candle before the moment when transaction should be executed
  
### Observed Results:

  * What happened?
 Observed the stop loss call that seems to be relevant to the data which is valid before the transaction execution.
  * What did you expect to happen?
 Stop loss should be executed obviously only after transaction execution.

**Comment by xmatthias** [comment]
Are you please able to provide us with a minimal reproduction?
basically a simple strategy that shows this problem, as well as the command used  (so we know the timerange, ...) to highlight the problem.

> request a transaction that will be executed in the middle of the candle

i'm currently not sure how this should be possible - entries are only checked against the long timeframe (all signals are on the long timeframe only, so using the detail-timeframe would accomplish nothing other than slowing down backtesting).

As such, an entry will either fill at open - or at the next "long" candle (if price is used outside of the initial candle).

**Comment by mciepluc** [comment]
Hi @xmatthias,

OK So you are saying that entries are checked only at the "long" candle open? What if I have trade at price that is reached in the middle of the "long" candle? Then the trade will be placed at the open of the next one? And using detail-timeframe is not able to fix this behavior?

**Comment by xmatthias** [comment]
I'm not sure what the purpose of this would be if i'm honest.

It'd be slower - and you'd get "by minute" fills (fill time is not kept in backtesting though) - so your trade will still be at 08:00.

if you're trading on 1h candles (your main timeframe is 1h) - then your trade duration will not be <1h.
That'd be guessing at best - as you'd never have the information to know if the next 1h candle is a green or a red one - not without sub-hour data.


So if the trade fills at 08:00, or at 08:45 - it'll make little difference.
it'll either hit stoploss within this candle, or it doesn't (backtesting will make it at 08:00 - as price was between high and low).

So assuming you have an entry at 08:00 - with a price between high/low of the 08:00-09:00 candle - it'd fill at 08:00 - and exit processing (in detail timeframe) will commence at 08:01 (actually at 08:00 - but for clarity's sake ...).

---

It might be an improvement to change behavior - but i'm not sure its worth the additional complexity - and potentially the additional performance penalty this would imply.

it might even make gaming backtesting easier - but i'm not sure it'd really improve accuracy (alignment with dry-runs) by a relevant margin.

**Comment by mciepluc** [comment]
Hi @xmatthias ,

You can easilly imagine situation when at a single 1 hour candle you enter the trade and the stoploss is reached. My expectation was that it can be handled using the timeframe-detail option. That is expected to provide higher resolution in trades. If that's not the case - what is the advantage of using timeframe-detail? I understand that under current implementation assumptions we must change the "long" interval to minutes and adjust the strategy to it to ensure the resolution requirement is satisfied.

**Comment by xmatthias** [comment]
I don't see how that would change behavior, really (not if you think about it carefully).

Assuming you place a long order on a 1h candle with a range of 90-110 - with a price of 100.

if you fill the order at :00 - or at :30 won't really change the outcome (not for the initial candle).
the fill price will still be 100 - and the low point has to be after the :30 point (otherwise it would have filled earlier - the moment price is ==100.

therefore stoploss calcs will work just fine (we know the open price - and we know the candle's low point is reached after the trade fills.

if the stoploss is within the candle's low point - it'll trigger either way (and will fill at the stop price).

As said above - it can be a future enhancement (not something i'll have time to look at anytime soon) - but it's not diminishing the value of timeframe-detail - as it will not make a real-world difference.

**Comment by maciekczwa** [comment]
It is making a significant order when using custom_entry_price.
Imagine this situation
Candle 1h opens on 110, falls down to 100, then falls back to 90
9:00 - 110
9:30 - 100
9:59 - 90

You used custom_entry_price with value 100 and position is long.

You are using following custom_stop_loss 
```
    stop_loss = 0.05
    use_custom_stoploss = False
    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                        current_profit: float, **kwargs) -> float:
        if current_profit > 0.05
            return 0.05  # trailing stop
        return 1  # default SL
```

The custom stop loss is executed from date time 9:00
So the first execution is at current_rate 110 so the current_profit is 0.10 the trailing stop in custom_stop_loss is activated.
The price falls down to 105 the trade is finished with profit 5%

In reality the order is filled at 100, and the price keeps falling down until it hits default stop loss 5%
So the trade ends on 95 with 5% loss.

I also checked the sitution with default trailing_stop_loss and there is a similar bug.

Everything is OK, if we are not using custom_entry_price and the order is filled ~ at the open of the candle.
